### ISSUE TITLE
Collection Variant Resolver: Performance Issues

### ISSUE THREAD

**Issue by xrev87** [vendure-ecommerce/vendure#1718]
**Describe the bug**
The variant field resolver is causing us memory and blocking issues, on brief inspection it looks like some queries are returning a lot of rows, this could be related to some queries are still unnecessarily joining on collections despite being the result of the select in patch.

**Environment (please complete the following information):**
 - @vendure/core version: latest stable
 - Nodejs version: 16
 - Database (mysql/postgres etc): postgres

**Additional context**

Here's the output of getQuery() - there's no limit, even though one is specified when calling the query via options on the field resolver (not the outer collection options)

```sql
SELECT *
FROM "product_variant" "productvariant"
         LEFT JOIN "product_variant_channels_channel" "productvariant_productvariant__channels"
                   ON "productvariant_productvariant__channels"."productVariantId" = "productvariant"."id"
         LEFT JOIN "channel" "productvariant__channels"
                   ON "productvariant__channels"."id" = "productvariant_productvariant__channels"."channelId"
         LEFT JOIN "product_variant_translation" "productvariant_translations"
                   ON "productvariant_translations"."baseId" = "productvariant"."id"
         LEFT JOIN "collection_product_variants_product_variant" "collection_productvariant"
                   ON "collection_productvariant"."productVariantId" = "productvariant"."id"
         LEFT JOIN "collection" "collection" ON "collection"."id" = "collection_productvariant"."collectionId"
         LEFT JOIN "product" "product" ON "product"."id" = "productvariant"."productId"
WHERE 1 = 1
  AND "productvariant__channels"."id" = :channelId
  AND "product"."deletedAt" IS NULL
  AND "productvariant"."deletedAt" IS NULL
  AND "collection"."id" = :collectionId
```

Obviously that's quite different from what actually gets executed via getManyAndCount() - which looks to be a series of select (distinct id) from and select from where id in (...) and the get distinct id queries do look to have the limit passed though

```sql
SELECT *
FROM "product_variant" "productvariant"
         LEFT JOIN "product_variant_channels_channel" "productvariant_productvariant__channels"
                   ON "productvariant_productvariant__channels"."productVariantId" = "productvariant"."id"
         LEFT JOIN "channel" "productvariant__channels"
                   ON "productvariant__channels"."id" = "productvariant_productvariant__channels"."channelId"
         LEFT JOIN "product_variant_translation" "productvariant_translations"
                   ON "productvariant_translations"."baseId" = "productvariant"."id"
         LEFT JOIN "collection_product_variants_product_variant" "collection_productvariant"
                   ON "collection_productvariant"."productVariantId" = "productvariant"."id"
         LEFT JOIN "collection" "collection" ON "collection"."id" = "collection_productvariant"."collectionId"
         LEFT JOIN "product" "product" ON "product"."id" = "productvariant"."productId"
WHERE (1 = 1 AND "productvariant__channels"."id" = $1 AND "product"."deletedAt" IS NULL AND
       "productvariant"."deletedAt" IS NULL AND "collection"."id" = $2)
  AND ("productvariant"."id" IN (88))
```

I'm thinking it shouldn't be joining back on collections there because we don't need them at this point due to having already retrieved the distinct ids of the productVariants in the collection

Here's the corresponding distinct id query:
```sql
SELECT DISTINCT "distinctAlias"."productvariant_id" as "ids_productvariant_id"
FROM (SELECT *
      FROM "product_variant" "productvariant"
               LEFT JOIN "product_variant_channels_channel" "productvariant_productvariant__channels"
                         ON "productvariant_productvariant__channels"."productVariantId" = "productvariant"."id"
               LEFT JOIN "channel" "productvariant__channels"
                         ON "productvariant__channels"."id" = "productvariant_productvariant__channels"."channelId"
               LEFT JOIN "product_variant_translation" "productvariant_translations"
                         ON "productvariant_translations"."baseId" = "productvariant"."id"
               LEFT JOIN "collection_product_variants_product_variant" "collection_productvariant"
                         ON "collection_productvariant"."productVariantId" = "productvariant"."id"
               LEFT JOIN "collection" "collection" ON "collection"."id" = "collection_productvariant"."collectionId"
               LEFT JOIN "product" "product" ON "product"."id" = "productvariant"."productId"
      WHERE 1 = 1
        AND "productvariant__channels"."id" = $1
        AND "product"."deletedAt" IS NULL
        AND "productvariant"."deletedAt" IS NULL
        AND "collection"."id" = $2) "distinctAlias"
ORDER BY "productvariant_id" ASC
LIMIT 1 
```

**Comment by michaelbromley** [comment]
Additional context on the type of dataset involved:

> 51,618 variants in total (quite a few disabled), max variants in any one collection is 6,197 (these are shop-all-parent type collections), max of 6 translations per variant and that's all fairly evenly distributed across 3 channels (excluding default).

**Comment by michaelbromley** [comment]
I just tested this out on a local dataset that I have, which is 

- ~11k ProductVariants total
- on a Collection with 5k variants
- single channel
- single language

 The following query executes in 50 - 100ms:
```gql
{
  collection(slug: "paint") {
    id
    name
    productVariants(options: { take: 100 }) {
      totalItems
      items {
        id
        name
      }
    }
  }
}
```
so I suspect that perhaps there is an issue caused by the multiple translations and/or channels which is leading to a combinatorial explosion of data being selected at the DB level.

First step to debugging this would be to construct a dataset with say 5k variants, each in 5 languages and 3 channels, and have them all added to a single Collection.

Let me know if you have time to write such a script.

**Comment by xrev87** [comment]
@michaelbromley  actually I was completely off base - the above queries weren't generated using the production data, merely a test dataset.

I was able to dump the prod database and do a slow query log locally and managed to catch this little number - which makes much more sense in terms of what I'm seeing:

```sql
SELECT *
FROM "collection" "collection"
         LEFT JOIN "collection" "collection__children" ON "collection__children"."parentId" = "collection"."id"
         LEFT JOIN "collection" "collection__children__parent"
                   ON "collection__children__parent"."id" = "collection__children"."parentId"
         LEFT JOIN "collection_translation" "collection__children__parent_translations"
                   ON "collection__children__parent_translations"."baseId" = "collection__children__parent"."id"
         LEFT JOIN "collection_product_variants_product_variant" "collection__children_collection__children__productVariants"
                   ON "collection__children_collection__children__productVariants"."collectionId" = "collection__children"."id"
         LEFT JOIN "product_variant" "collection__children__productVariants"
                   ON "collection__children__productVariants"."id" =
                      "collection__children_collection__children__productVariants"."productVariantId"
         LEFT JOIN "product_variant_translation" "collection__children__productVariants_translations"
                   ON "collection__children__productVariants_translations"."baseId" =
                      "collection__children__productVariants"."id"
         LEFT JOIN "collection_translation" "collection__children_translations"
                   ON "collection__children_translations"."baseId" = "collection__children"."id"
WHERE "collection"."id" IN  ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
```

Notice how product variants is unnecessarily getting joined there to facillitate child collections - it's completely forgoing the limit value set at a field resolver level.

**Comment by michaelbromley** [comment]
@xrev87 can you provide the GraphQL query that is causing that SQL query?

**Comment by xrev87** [comment]
> @xrev87 can you provide the GraphQL query that is causing that SQL query?

Sure:

```graphql
fragment BreadcrumbsProps on Collection {
  breadcrumbs {
    id
    name
    slug
    __typename
  }
  __typename
}

fragment CollectionProps on Collection {
  createdAt
  description
  id
  isPrivate
  name
  position
  slug
  productVariants(options: { take: 1 }) {
    totalItems
    __typename
  }
  __typename
}

fragment CustomFieldsProps on CollectionCustomFields {
  Featured
  __typename
}

fragment ChildrenProps on Collection {
  children {
    ...CollectionProps
    ...BreadcrumbsProps
    customFields {
      ...CustomFieldsProps
      __typename
    }
    __typename
  }
  __typename
}

fragment ParentProps on Collection {
  parent {
    ...CollectionProps
    __typename
  }
  __typename
}

query Categories($options: CollectionListOptions) {
  categories(options: $options) {
    items {
      ...CollectionProps
      ...BreadcrumbsProps
      ...ParentProps
      ...ChildrenProps
      customFields {
        ...CustomFieldsProps
        __typename
      }
      __typename
    }
    totalItems
    __typename
  }
}

// Options: sort: { position: ASC }
```

**Comment by xrev87** [comment]
> @xrev87 can you provide the GraphQL query that is causing that SQL query?

Sure:
```graphql
fragment BreadcrumbsProps on Collection {
  breadcrumbs {
    id
    name
    slug
    __typename
  }
  __typename
}

fragment CollectionProps on Collection {
  createdAt
  description
  id
  isPrivate
  name
  position
  slug
  productVariants(options: { take: 1 }) {
    totalItems
    __typename
  }
  __typename
}

fragment CustomFieldsProps on CollectionCustomFields {
  Featured
  __typename
}

fragment ChildrenProps on Collection {
  children {
    ...CollectionProps
    ...BreadcrumbsProps
    customFields {
      ...CustomFieldsProps
      __typename
    }
    __typename
  }
  __typename
}

fragment ParentProps on Collection {
  parent {
    ...CollectionProps
    __typename
  }
  __typename
}

query Collections($options: CollectionListOptions) {
  collections(options: $options) {
    items {
      ...CollectionProps
      ...BreadcrumbsProps
      ...ParentProps
      ...ChildrenProps
      customFields {
        ...CustomFieldsProps
        __typename
      }
      __typename
    }
    totalItems
    __typename
  }
}

// Options: sort: { position: ASC }
```

**Comment by michaelbromley** [comment]
Thanks, I think I discovered the root cause. This is quite a heavy query anyway due to the unbounded recursive nature of parent/children selection. But I have made a change that in my testing dataset reduces the query time from ~15s to ~3s (did not measure memory consumption).

Will put this fix in the upcoming v1.7.
